src/wp-includes/class-wpdb.php

```diff
@@ -1475,20 +1528,140 @@ public function prepare( $query, ...$args ) {
+		 * then 3 additional values per placeholder.
+		 */
+		$placeholder_count = ( ( $split_query_count - 1 ) / 3 );
+
+		// If args were passed as an array, as in vsprintf(), move them up.
+		$passed_as_array = ( isset( $args[0] ) && is_array( $args[0] ) && 1 === count( $args ) );
+		if ( $passed_as_array ) {
+			$args = $args[0];
+		}
 
-		// Count the number of valid placeholders in the query.
-		$placeholders = preg_match_all( "/(^|[^%]|(%%)+)%($allowed_format)?[sdF]/", $query, $matches );
+		$new_query       = '';
+		$key             = 2; // Keys 0 and 1 in $split_query contain values before the first placeholder.
+		$arg_id          = 0;
+		$arg_identifiers = array();
+		$arg_strings     = array();
+
+		while ( $key < $split_query_count ) {
+			$placeholder = $split_query[ $key ];
+
+			$format = substr( $placeholder, 1, -1 );
+			$type   = substr( $placeholder, -1 );
+
+			if ( 'f' === $type ) { // Force floats to be locale-unaware.
+				$type        = 'F';
+				$placeholder = '%' . $format . $type;
+			}
+
+			if ( 'i' === $type ) {
+				$placeholder = '`%' . $format . 's`';
+				// Using a simple strpos() due to previous checking (e.g. $allowed_format).
+				$argnum_pos = strpos( $format, '$' );
+
+				if ( false !== $argnum_pos ) {
+					// sprintf() argnum starts at 1, $arg_id from 0.
+					$arg_identifiers[] = ( intval( substr( $format, 0, $argnum_pos ) ) - 1 );
+				} else {
+					$arg_identifiers[] = $arg_id;
+				}
+			} elseif ( 'd' !== $type && 'F' !== $type ) {
+				/*
+				 * i.e. ( 's' === $type ), where 'd' and 'F' keeps $placeholder unchanged,
+				 * and we ensure string escaping is used as a safe default (e.g. even if 'x').
+				 */
+				$argnum_pos = strpos( $format, '$' );
+
+				if ( false !== $argnum_pos ) {
+					$arg_strings[] = ( intval( substr( $format, 0, $argnum_pos ) ) - 1 );
+				} else {
+					$arg_strings[] = $arg_id;
+				}
+
+				// Unquoted strings for backward compatibility (dangerous).
+				// First, "numbered or formatted string placeholders (eg, %1$s, %5s)".
+				// Second, if "%s" has a "%" before it, even if it's unrelated (e.g. "LIKE '%%%s%%'").
+				if ( true !== $this->allow_unsafe_unquoted_parameters || ( '' === $format && ! str_ends_with( $split_query[ $key - 1 ], '%' ) ) ) {
+					$placeholder = "'%" . $format . "s'";
+				}
+			}
+
+			// Glue (-2), any leading characters (-1), then the new $placeholder.
+			$new_query .= $split_query[ $key - 2 ] . $split_query[ $key - 1 ] . $placeholder;
+
+			$key += 3;
+			$arg_id++;
+		}
+
+		// Replace $query; and add remaining $query characters, or index 0 if there were no placeholders.
+		$query = $new_query . $split_query[ $key - 2 ];
+
+		$dual_use = array_intersect( $arg_identifiers, $arg_strings );
+
+		if ( count( $dual_use ) > 0 ) {
+			wp_load_translations_early();
+
+			$used_placeholders = array();
+
+			$key    = 2;
+			$arg_id = 0;
+			while ( $key < $split_query_count ) { // Parse again (only used when there is an error).
```

Comment:
```suggestion
			// Parse again (only used when there is an error).
			while ( $key < $split_query_count ) {
```
Same as above, do you mind if we move this inline comment above the block?

Reply:
Done